go-controller/pkg/node/gateway_iptables.go

```diff
@@ -355,14 +416,26 @@ func getGatewayIPTRules(service *kapi.Service, hasLocalHostEndpoint bool) []iptR
 				klog.Errorf("Skipping service: %s, invalid service port %v", svcPort.Name, err)
 				continue
 			}
-			if !hasLocalHostEndpoint {
+			if svcTypeIsETPLocal && svcHasLocalHostNetEndPnt {
+				// case1 (see function description for details)
+				// Port redirect host -> Nodeport -> host traffic directly to endpoint
 				for _, clusterIP := range clusterIPs {
-					rules = append(rules, getNodePortIPTRules(svcPort, clusterIP, svcPort.Port)...)
+					rules = append(rules, getNodePortETPLocalIPTRules(svcPort, clusterIP, int32(svcPort.TargetPort.IntValue()), false)...)
 				}
+			} else if svcTypeIsETPLocal && !svcHasLocalHostNetEndPnt && config.Gateway.Mode == config.GatewayModeLocal {
+				// case2 (see function description for details)
+				// will hit this only if 1)gatewaymode=local & 2)svc.externaltrafficpolicy=local & 3)we don't have
+				// any local-hostnetworked endpoints for this service (this includes 3a) endpoints.Subset is empty or
+				// 3b) endpoints are host-networked but not local to this node or 3c) endpoints are ovn-podIPs).
+				//
+				// DNAT traffic to masqueradeIP instead of clusterIP.
+				rules = append(rules, getNodePortIPTRules(svcPort, "", svcPort.NodePort, true)...)
+				// add a skip SNAT rule to OVN-KUBE-SNAT-MGMTPORT to preserve sourceIP for etp=local traffic.
+				rules = append(rules, getNodePortETPLocalIPTRules(svcPort, "", int32(svcPort.TargetPort.IntValue()), true)...)
```

Comment:
this is confusing, you pass true here which is svcDoesNotHaveLocalHostNetEndPnt, but you are checking !svcHasLocalHostNetEndPnt. This is another reason to change the function signature to match and not use negation. Then just pass the variable to the function and do the logic there.

Reply:
ack, makes sense